File: gym_http_client.py
import requests
import six.moves.urllib.parse as urlparse
import json
import os

import logging
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

class Client(object):
    """
    Gym client to interface with gym_http_server
    """
    def __init__(self, remote_base):
        self.remote_base = remote_base
        self.session = requests.Session()
        self.session.headers.update({'Content-type': 'application/json'})

    def _parse_server_error_or_raise_for_status(self, resp):
        j = {}
        try:
            j = resp.json()
        except:
            # Most likely json parse failed because of network error, not server error (server
            # sends its errors in json). Don't let parse exception go up, but rather raise default
            # error.
            resp.raise_for_status()
        if resp.status_code != 200 and "message" in j:  # descriptive message from server side
            raise ServerError(message=j["message"], status_code=resp.status_code)
        resp.raise_for_status()
        return j

    def _post_request(self, route, data):
        url = urlparse.urljoin(self.remote_base, route)
        logger.info("POST {}\n{}".format(url, json.dumps(data)))
        resp = self.session.post(urlparse.urljoin(self.remote_base, route),
                            data=json.dumps(data))
        return self._parse_server_error_or_raise_for_status(resp)

    def _get_request(self, route):
        url = urlparse.urljoin(self.remote_base, route)
        logger.info("GET {}".format(url))
        resp = self.session.get(url)
        return self._parse_server_error_or_raise_for_status(resp)
        
    def env_create(self, env_id):
        route = '/v1/envs/'
        data = {'env_id': env_id}
        resp = self._post_request(route, data)
        instance_id = resp['instance_id']
        return instance_id

    def env_list_all(self):
        route = '/v1/envs/'
        resp = self._get_request(route)
        all_envs = resp['all_envs']
        return all_envs

    def env_reset(self, instance_id):
        route = '/v1/envs/{}/reset/'.format(instance_id)
        resp = self._post_request(route, None)
        observation = resp['observation']
        return observation

    def env_step(self, instance_id, action, render=False):
        route = '/v1/envs/{}/step/'.format(instance_id)
        data = {'action': action, 'render': render}
        resp = self._post_request(route, data)
        observation = resp['observation']
        reward = resp['reward']
        done = resp['done']
        info = resp['info']
        return [observation, reward, done, info]

    def env_action_space_info(self, instance_id):
        route = '/v1/envs/{}/action_space/'.format(instance_id)
        resp = self._get_request(route)
        info = resp['info']
        return info

    def env_action_space_sample(self, instance_id):
        route = '/v1/envs/{}/action_space/sample'.format(instance_id)
        resp = self._get_request(route)
        action = resp['action']
        return action

    def env_action_space_contains(self, instance_id, x):
        route = '/v1/envs/{}/action_space/contains/{}'.format(instance_id, x)
        resp = self._get_request(route)
        member = resp['member']
        return member

    def env_observation_space_info(self, instance_id):
        route = '/v1/envs/{}/observation_space/'.format(instance_id)
        resp = self._get_request(route)
        info = resp['info']
        return info

    def env_observation_space_contains(self, instance_id, params):
        route = '/v1/envs/{}/observation_space/contains'.format(instance_id)
        resp = self._post_request(route, params)
        member = resp['member']
        return member

    def env_monitor_start(self, instance_id, directory,
                              force=False, resume=False, video_callable=False):
        route = '/v1/envs/{}/monitor/start/'.format(instance_id)
        data = {'directory': directory,
                'force': force,
                'resume': resume,
                'video_callable': video_callable}
        self._post_request(route, data)

    def env_monitor_close(self, instance_id):
        route = '/v1/envs/{}/monitor/close/'.format(instance_id)
        self._post_request(route, None)

    def env_close(self, instance_id):
        route = '/v1/envs/{}/close/'.format(instance_id)
        self._post_request(route, None)

    def upload(self, training_dir, algorithm_id=None, api_key=None):
        if not api_key:
            api_key = os.environ.get('OPENAI_GYM_API_KEY')

        route = '/v1/upload/'
        data = {'training_dir': training_dir,
                'algorithm_id': algorithm_id,
                'api_key': api_key}
        self._post_request(route, data)

    def shutdown_server(self):
        route = '/v1/shutdown/'
        self._post_request(route, None)

class ServerError(Exception):
    def __init__(self, message, status_code=None):
        Exception.__init__(self)
        self.message = message
        if status_code is not None:
            self.status_code = status_code

if __name__ == '__main__':
    remote_base = 'http://127.0.0.1:5000'
    client = Client(remote_base)

    # Create environment
    env_id = 'CartPole-v0'
    instance_id = client.env_create(env_id)

    # Check properties
    all_envs = client.env_list_all()
    action_info = client.env_action_space_info(instance_id)
    obs_info = client.env_observation_space_info(instance_id)

    # Run a single step
    client.env_monitor_start(instance_id, directory='tmp', force=True)
    init_obs = client.env_reset(instance_id)
    [observation, reward, done, info] = client.env_step(instance_id, 1, True)
    client.env_monitor_close(instance_id)
    client.upload(training_dir='tmp')

File: binding-lua/gym_http_client.lua
local HttpClient = require("httpclient")
local json = require("dkjson")
local os = require 'os'

local GymClient = {}
local m = {}

function m.new(remote_base)
   local self = {}
   self.remote_base = remote_base
   self.http = HttpClient.new()
   setmetatable(self, {__index = GymClient})
   return self
end

function GymClient:parse_server_error_or_raise_for_status(resp)
   local resp_data, pos, err = {}
   if resp.err then
      err = resp.err
      -- print('Response error: ' .. err)
   else
      if resp.code ~= 200 then
         err = resp.status_line
         -- Descriptive message from the server side
         -- print('Response: ' .. err)
      end
      resp_data, pos, err = json.decode(resp.body)
   end
   return resp_data, pos, err
end

function GymClient:get_request(route)
   url = self.remote_base .. route
   options = {}
   options.content_type = 'application/json'
   resp = self.http:get(url, options)
   return self:parse_server_error_or_raise_for_status(resp)
end

function GymClient:post_request(route, req_data)
   url = self.remote_base .. route
   options = {}
   options.content_type = 'application/json'
   json_str = json.encode(req_data)
   resp = self.http:post(url, json_str, options)
   return self:parse_server_error_or_raise_for_status(resp)
end

function GymClient:env_create(env_id)
   route = '/v1/envs/'
   req_data = {env_id = env_id}
   resp_data = self:post_request(route, req_data)
   return resp_data['instance_id']
end

function GymClient:env_list_all()
   route = '/v1/envs/'
   resp_data = self:get_request(route)
   return resp_data['all_envs']
end

function GymClient:env_reset(instance_id)
   route = '/v1/envs/'..instance_id..'/reset/'
   resp_data = self:post_request(route, '')
   return resp_data['observation']
end

function GymClient:env_step(instance_id, action, render, video_callable)
   render = render or false
   video_callable = video_callable or false
   route = '/v1/envs/'..instance_id..'/step/'
   req_data = {action = action, render = render, video_callable = video_callable}
   resp_data = self:post_request(route, req_data)
   obs = resp_data['observation']
   reward = resp_data['reward']
   done = resp_data['done']
   info = resp_data['info']
   return obs, reward, done, info
end

function GymClient:env_action_space_info(instance_id)
   route = '/v1/envs/'..instance_id..'/action_space/'
   resp_data = self:get_request(route)
   return resp_data['info']
end

function GymClient:env_action_space_sample(instance_id)
   route = '/v1/envs/'..instance_id..'/action_space/sample'
   resp_data = self:get_request(route)
   action = resp_data['action']
   return action
end

function GymClient:env_action_space_contains(instance_id)
   route = '/v1/envs/'..instance_id..'/action_space/contains'
   resp_data = self:get_request(route)
   member = resp['member']
   return member
end

function GymClient:env_observation_space_info(instance_id)
   route = '/v1/envs/'..instance_id..'/observation_space/'
   resp_data = self:get_request(route)
   return resp_data['info']
end

function GymClient:env_monitor_start(instance_id, directory, force, resume, video_callable)
   if not force then force = false end
   if not resume then resume = false end
   req_data = {directory = directory,
            force = tostring(force),
            resume = tostring(resume),
            video_callable = video_callable}
   route = '/v1/envs/'..instance_id..'/monitor/start/'
   resp_data  = self:post_request(route, req_data)
end

function GymClient:env_monitor_close(instance_id)
   route = '/v1/envs/'..instance_id..'/monitor/close/'
   resp_data = self:post_request(route, '')
end

function GymClient:upload(training_dir, algorithm_id, api_key)
   if not api_key then
     api_key = os.getenv('OPENAI_GYM_API_KEY')
   end
   if not algorithm_id then algorithm_id = '' end
   req_data = {training_dir = training_dir,
            algorithm_id = algorithm_id,
            api_key = api_key}
   route = '/v1/upload/'
   resp = self:post_request(route, req_data)
   return resp
end

function GymClient:shutdown_server()
   route = '/v1/shutdown/'
   self:post_request(route, '')
end

return m